design-proposals/vm-backend-storage.md

```diff
@@ -0,0 +1,73 @@
+# Overview
+Today, anything that happens on the host side of VMIs (i.e. virt-launcher) is transient.  
+Any change a VMI makes outside of its own file-system is lost after a shutdown.  
+Such changes include EFI firmware settings and information stored inside the TPM.  
+This design provides a way to persist small VM-specific files used by virt-launcher.
+
+## Motivation
+Though we now support exposing an emulated TPM device to guests (vTPM), using the device in any meaningful way would require persisting data across reboots.  
+As we do not currently have a way to do that, all data stored by the VMI into the vTPM is lost after a shutdown (a soft reboot keeps the same virt-launcher, avoiding losses).
+
+## Goals
+- Being able to persist modifications of various emulated devices across reboots for a given VM
+- Enable live migration of VMs that use backend storage
+- Avoid consuming large amounts of storage
+- Prevent virt-launcher for VM X from seeing data that belongs to VM Y
+
+## Non Goals
+- Persisting everything. Some information that would usually persist on a non-ephemeral host just don't matter to us.
+- Supporting user modifications of the backend storage. Users whishing to inject their own files should do so at their own risk.
+
+## Definition of Users
+This feature is intended for end-users who wish to preserve the state of some emulated device across reboots.
+
+## User Stories
+As a VM user, I want to be able to store the encryption key of my VMs partitions inside the vTPM (example: BitLocker)
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+The storage used for this would be 1 PVC per VM, and would be created on first start of a VM that needs backend storage.  
+This PVC would be RWO, about 2 MiB in size (still TBD), and use [csi-hostpath](https://github.com/kubevirt/hostpath-provisioner).  
+The PVC would be added to the virt-launcher compute container as a volume, but never assigned to the libvirt domain (the guest VM).  
+The PVC would be mounted in a directory at the root of the filesystem. Appropriate symbolic links will be created before starting the libvirt domain.
+The PVC would be owned-by the VM, and therefore will get deleted when its associated VM is deleted.
+
+## API Examples
+Currently, the TPM device API doesn't have any option.  
```

Comment:
I think we should state here that there will be no dedicated API field for this and features such as persistent TPM will transparently use the storage.

Edit: Will there be kubevirt option to provide the storage class for the storage?

Reply:
Right, there will be a `persistent` bool, but no explicit "enable backend storage" option.
Adding a KubeVirt option to provide the storage class is definitely a possibility.